vllm/worker/worker_base.py

```diff
@@ -558,10 +558,34 @@ def init_worker(self, all_kwargs: List[Dict[str, Any]]) -> None:
             worker_class = resolve_obj_by_qualname(
                 self.vllm_config.parallel_config.worker_cls)
         else:
+            logger.warning(
+                "passing worker_cls as a class object is strongly deprecated,"
+                " as the serialization of class objects can be tricky and"
+                " error-prone. To be safe, please keep the class in a separate"
+                " module and pass the qualified name of the class as a string."
+            )
             assert isinstance(self.vllm_config.parallel_config.worker_cls,
                               bytes)
             worker_class = cloudpickle.loads(
                 self.vllm_config.parallel_config.worker_cls)
+        if self.vllm_config.parallel_config.worker_adapter_cls:
+            worker_adapter_class = resolve_obj_by_qualname(
+                self.vllm_config.parallel_config.worker_adapter_cls)
+            logger.info(
+                "Injecting %s into %s for extended collective_rpc call",
+                worker_adapter_class, worker_class)
+            if worker_adapter_class not in worker_class.__bases__:
+                # check any conflicts between worker and worker_adapter
+                for attr in dir(worker_adapter_class):
+                    if attr.startswith("__"):
+                        continue
+                    assert not hasattr(worker_class, attr), (
+                        f"Worker class {worker_class} already has an attribute"
+                        f" {attr}, which conflicts with the worker"
+                        f" adapter class {worker_adapter_class}.")
+                # dynamically inherit the worker adapter class
+                worker_class.__bases__ = worker_class.__bases__ + (
+                    worker_adapter_class, )
```

Comment:
this is wild

Reply:
honestly pretty cool that you can do this

Reply:
any suggestion to make it better? the problem is, the worker class is dynamically determined based on v1 or v0.

Reply:
Do you think we'd be able to simplify this once we move off of V0 completely?

Reply:
I don't think so, this is not just for v0/v1, but also for different platforms. for example, if an rlhf framework wants to support both cuda and rocm, they can use a similar mixin class for both, without knowing what is the underlying worker class.